torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -253,6 +293,13 @@ def _share_state_and_init_handle_attrs(
         handle = fsdp_state._handle
         if handle:
             handle.init_flat_param_attributes()
+    # Ensure that all unshard streams wait on the default computation stream
```

Comment:
Wait sorry, I think this needs to be done in the root pre-forward so that we call this every iteration. I think `_share_state_and_init_handle_attrs` should only run once.

Here is where the wait stream happens in `_root_pre_forward`:
https://github.com/pytorch/pytorch/blob/2b2b3c453660bf57543dc3420007c5ccf1317809/torch/distributed/fsdp/_runtime_utils.py#L646-L650

I think ideally, we should refactor `_wait_for_computation_stream()` to take in the root state and inside the function, we get all unshard streams in all FSDP states, the reason being that there is still only one pre-unshard stream shared across all FSDP states.

Reply:
Ah I see, yes this makes sense to do. I thought `_lazy_init`, and by extension, `_share_state_and_init_handle_attrs`, would run in every call to `_root_pre_forward`. But didn't realize it's a no-op after the first call to forward. I'll get this change in now.